engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java

```diff
@@ -534,4 +534,25 @@ private Set<Long> getBackoffJobKey() {
         (key, value) -> backoffJobKeys.add(key.second().inner().getValue()));
     return backoffJobKeys;
   }
+
+  private boolean isValidForRestore(final long key, final JobRecordValue jobRecord) {
+    if (jobRecord == null) {
+      return false;
+    }
+    final var job = jobRecord.getRecord();
+    return job.getRecurringTime() > -1 && job.getRetries() > 0 && isJobActivatable(key, job);
+  }
+
+  private boolean isJobActivatable(final long key, final JobRecord job) {
+    final var type = job.getTypeBuffer();
+    final var tenantId = job.getTenantId();
+    EnsureUtil.ensureNotNullOrEmpty("type", job.getTypeBuffer());
+    EnsureUtil.ensureNotNullOrEmpty("tenantId", job.getTenantId());
+
+    jobTypeKey.wrapBuffer(type);
+    jobKey.wrapLong(key);
+    tenantIdKey.wrapString(tenantId);
+
+    return activatableColumnFamily.get(tenantAwareTypeJobKey) != null;
```

Comment:
:question: @nicpuppa why are we checking if the job is activatable?

If the job is already in the `activatableColumnFamily` that means that it hasn't been `FAILED`, and it shouldn't be in the `backoffColumnFamily`.

So if we can find the job in the `activatableColumnFamily`, then we shouldn't put it in the `backoffColumnFamily`.

Am I missing some steps in my logic?

Reply:
Probably I misunderstood, but in the `backoffColumnFamily` you should find only activatable jobs. When a job is failed, and have retries, is still available in the `activatableColumnFamily`

Reply:
There are 3 scenarios once a Job is `FAILED` (see code [here](https://github.com/camunda/zeebe/blob/86db453f1aef24e2fb02e7a739dda12a12b1997c/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java#L193-L205)):
1. If `retires <= 0`
   - `JobState` is set to `FAILED`.
   - `Job` is removed from the `activatableColumnFamily`
   - Incident is added for this Job.
2. If `retries > 0` and `retryBackoff > 0`
   - `JobState` is set to `FAILED`.
   - `Job` was already removed from the `activatableColumnFamily` [when the job was activated](https://github.com/camunda/zeebe/blob/86db453f1aef24e2fb02e7a739dda12a12b1997c/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java#L132). So it's currently not there.
   - `Job` is added to the `backoffColumnFamily` since we want to delay the next job activation until some time has passed.
3. If `retries > 0` and `retryBackoff <= 0`
   - `JobState` is set to `ACTIVATABLE`.
   - `Job` is added to the `activatableColumnFamily` since we want it to be immediately available for activation.

For this migration, we need to fix case [2], so:
1. A job should have `retries > 0` :heavy_check_mark: 
2. A job should have `recurringTime > -1` :heavy_check_mark: 
3. A job is not activatable, i.e. `!isJobActivatable(key, job) :x:

Since we already do [`getFailedJobKeys()`](https://github.com/camunda/zeebe/pull/16047/files#diff-ef73a79bfa12b3c9cc862cff7aefcdc63b8cb9efaa732c6de49b842915017b20R281), i.e. [jobs in state `FAILED`](https://github.com/camunda/zeebe/pull/16047/files#diff-ef73a79bfa12b3c9cc862cff7aefcdc63b8cb9efaa732c6de49b842915017b20R524), finding the job in the `activatableColumnFamily` would indicate a bug since a job in that CF must be in the `ACTIVATABLE` state.

Reply:
And I think we might have another bug, since the `JobFailProcessor` allows to `FAIL` an `ACTIVATABLE` job ([src](https://github.com/camunda/zeebe/blob/2e22958504385ec243ffd4e932c2283b3ad5104f/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java#L72)).

When an `ACTIVATABLE` job is failed **with backoff**, it isn't removed from the `activatableColumnFamily` when it's marked as `FAILED` and placed in the `backoffColumnFamily`. This would make it possible for a Job to be activated earlier than the `recurringTime`. This defeats the purpose of `backoff`.

Reply:
Oh thanks for clarify this, the current behaviour mislead me 👍